### PR TITLE
Fix cask source file path loading issues

### LINKED ISSUE #17226: Casks can fail to load when source file path is nil
### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  llvm@11
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.2.20-104-gdfbf269
ORIGIN: https://github.com/Homebrew/brew
HEAD: dfbf26910d514220e6453f6efdd5d9fb4d36c145
Last commit: 11 hours ago
Core tap HEAD: 5c9c84e8d43c9ab364371cef3f9b7473d91d954b
Core tap last commit: 2 hours ago
Core tap JSON: 04 May 01:07 UTC
Core cask tap JSON: 04 May 01:07 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_AUTOREMOVE: set
HOMEBREW_CASK_OPTS: []
HOMEBREW_DEVELOPER: set
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.kTNdVTmWzn/org.macosforge.xquartz:0
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: quad-core 64-bit ivybridge
Clang: 12.0.0 build 1200
Git: 2.24.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: 12.4
```

### What were you trying to do (and why)?

I was trying to reproduce this cask loading issue: https://github.com/orgs/Homebrew/discussions/4910#discussioncomment-8750211

CC: @johnmcdowell

### What happened (include all command output)?

```console
$ brew bump --cask iterm2 --verbose --debug
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/bin/bundle clean
/usr/local/Homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::FromInstalledPathLoader): loading iterm2
Error: Parameter 'ref': Expected type T.any(Cask::Cask, Pathname, String, URI::Generic), got type NilClass
Caller: /usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:564
Definition: /usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:235
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/configuration.rb:296:in `call_validation_error_handler_default'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/configuration.rb:303:in `call_validation_error_handler'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:300:in `report_error'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:136:in `block in validate_call_skip_block_type'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/signature.rb:213:in `each_args_value_type'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:133:in `validate_call_skip_block_type'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:109:in `block in create_validator_slow_skip_block_type'
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:564:in `block in for'
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:563:in `each'
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:563:in `for'
/usr/local/Homebrew/Library/Homebrew/cask/cask_loader.rb:514:in `load'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:356:in `block (2 levels) in retrieve_versions_by_arch'
/usr/local/Homebrew/Library/Homebrew/simulate_system.rb:27:in `with'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:348:in `block in retrieve_versions_by_arch'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:347:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:347:in `retrieve_versions_by_arch'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:443:in `retrieve_and_display_info_and_open_pr'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:159:in `block in handle_formula_and_casks'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:145:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:145:in `each_with_index'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:145:in `handle_formula_and_casks'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:103:in `block in run'
/usr/local/Homebrew/Library/Homebrew/extend/kernel.rb:520:in `with_env'
/usr/local/Homebrew/Library/Homebrew/api.rb:196:in `with_no_api_env'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/bump.rb:68:in `run'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `bind_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/call_validation.rb:270:in `validate_call'
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/3.1.0/gems/sorbet-runtime-0.5.11368/lib/types/private/methods/_methods.rb:277:in `block in _on_method_added'
/usr/local/Homebrew/Library/Homebrew/brew.rb:92:in `<main>'
```

### What did you expect to happen?

It shouldn't cause a type error.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. Install a cask from `homebrew/cask`.
2. Untap the `homebrew/cask` repo.
3. Run `brew bump --cask INSTALLED_CASK_NAME`.
```

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Fixes #17226

There are two big changes here. Both have to do with how we want to load casks in different scenarios. One also is related to formulae.

1. Prevent loading casks & formulae outside of taps for specific commands.

There are certain commands like `bump`, `bump-*-pr`, `livecheck` and `audit` where it really makes no sense to try and run things if the specified formulae or cask is not in a tap. A new `#to_formulae_and_casks_with_taps` method was added to the `CLI::NamedArgs` class to allow us to easily grab and validate formulae and casks from named arguments.

2. Always load the source file path when loading casks with the path loader.

There was an edge case where all JSON cask files were being loaded without setting the source file path because most of the work was handed off to the API loader where that normally would make more sense. Now we set that when calling the API loader which solves the problem. This improves the user experience of people using the `--cache` and `fetch` commands in certain edge cases. Hopefully it makes the user experience a bit more consistent.

A regression test was added for this point.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/bump.rb [hunk 1]
```diff
@@ -78,13 +78,7 @@ def run
...
             formulae + casks
           elsif args.named.present?
-            if args.formula?
-              args.named.to_formulae
-            elsif args.cask?
-              args.named.to_casks
-            else
-              args.named.to_formulae_and_casks
-            end
+            args.named.to_formulae_and_casks_with_taps
```
Comment (apainintheneck) [inline Library/Homebrew/dev-cmd/bump.rb:81]: Before this changes this was equivalent to `#to_formulae_and_casks` since the `--cask` and `--formula` options get passed to that method automatically by the parser.

File: Library/Homebrew/cli/named_args.rb [hunk 2]
```diff
@@ -103,6 +103,20 @@ def to_formulae_to_casks(only: parent&.only_formula_or_cask, method: nil)
...
+          case formula_or_cask
+          when Formula
+            odie "Formula #{formula_or_cask.name} is not in a tap!" unless formula_or_cask.tap
+          when Cask::Cask
+            odie "Cask #{formula_or_cask.token} is not in a tap!" unless formula_or_cask.tap
+          else
+            raise ArgumentError, "Expected a formula or a cask: #{formula_or_cask}"
+          end
+        end
+      end
```
Comment (apainintheneck) [inline Library/Homebrew/cli/named_args.rb:118]: This message can maybe be improved by adding something about tapping the core taps locally. I'll have to think about it.

File: Library/Homebrew/cli/named_args.rb [hunk 3]
```diff
@@ -103,6 +103,20 @@ def to_formulae_to_casks(only: parent&.only_formula_or_cask, method: nil)
...
       end
 
+      # Returns formulae and casks after validating that a tap is present for each of them.
+      def to_formulae_and_casks_with_taps
+        to_formulae_and_casks.each do |formula_or_cask|
+          case formula_or_cask
+          when Formula
+            odie "Formula #{formula_or_cask.name} is not in a tap!" unless formula_or_cask.tap
+          when Cask::Cask
+            odie "Cask #{formula_or_cask.token} is not in a tap!" unless formula_or_cask.tap
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/named_args.rb:113]: Might be better to gather all the missing formulae/casks outside of the loop and print a single error message. This avoids having e.g. `brew audit badformula1 badcask1 badformula2 badcask2` printing 4 different error messages but instead one good one.

File: Library/Homebrew/cli/named_args.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cli/named_args.rb:118]: > is not in any locally tapped taps. You may need to `brew tap` additional taps.

Given this is longer/shared: it pairs nicely with suggestion above.

**Review by MikeMcQuaid — APPROVED** [review]
Approach looks good, thanks @apainintheneck!

File: Library/Homebrew/cli/named_args.rb [hunk 4]
```diff
@@ -103,6 +103,34 @@ def to_formulae_to_casks(only: parent&.only_formula_or_cask, method: nil)
...
+        formulae_and_casks_with_taps, formulae_and_casks_without_taps =
+          to_formulae_and_casks.partition do |formula_or_cask|
+            case formula_or_cask
+            when Formula, Cask::Cask
+              formula_or_cask.tap&.installed?
+            else
+              # TODO: Refactor methods so that Sorbet can tell that this is unreachable.
+              raise ArgumentError, "Not a formula or cask: #{formula_or_cask.class}"
+            end
+          end
```
Comment (apainintheneck) [inline Library/Homebrew/cli/named_args.rb:117]:
```suggestion
          to_formulae_and_casks.partition { |formula_or_cask| formula_or_cask.tap&.installed? }
```

If Sorbet was smarter, it would look like this. I think that things will have to be refactored a bit in the named args class and I'm not sure if that's in scope here.

**Comment by apainintheneck** [comment]
The error message now looks like this:

```
Error: These formulae and casks are not in any locally installed taps!

- discord
- iterm2
- visual-studio-code
- zstd

You may need to run `brew tap` to install additional taps.
```

File: Library/Homebrew/cli/named_args.rb [hunk 5]
```diff
@@ -103,6 +103,34 @@ def to_formulae_to_casks(only: parent&.only_formula_or_cask, method: nil)
...
+        return formulae_and_casks_with_taps if formulae_and_casks_without_taps.blank?
+
+        types = []
+        types << "formulae" if formulae_and_casks_without_taps.any?(Formula)
+        types << "casks" if formulae_and_casks_without_taps.any?(Cask::Cask)
+
+        odie <<~ERROR
+          These #{types.join(" and ")} are not in any locally installed taps!
+
+          - #{formulae_and_casks_without_taps.sort_by(&:to_s).join("\n- ")}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/named_args.rb:128]:
```suggestion
            #{formulae_and_casks_without_taps.sort_by(&:to_s).join("\n  ")}
```

We don't tend to use `-` here just to make piping a bit easier.

**Review by MikeMcQuaid — APPROVED** [review]
Good to go once suggestion applied, thanks @apainintheneck!

File: Library/Homebrew/cli/named_args.rb (hunk 4 above)
Comment (carlocab) [reply Library/Homebrew/cli/named_args.rb:117]:
```suggestion
          to_formulae_and_casks.partition { |formula_or_cask| T.cast(formula_or_cask, T.any(Formula, Cask)).tap&.installed? }
```
doesn't work?

Comment (apainintheneck) [reply Library/Homebrew/cli/named_args.rb]: @carlocab Yep, that worked. Thanks!

File: Library/Homebrew/cli/named_args.rb (hunk 5 above)
Comment (apainintheneck) [reply Library/Homebrew/cli/named_args.rb]: The new error message looks like this.

```
Error: These formulae are not in any locally installed taps!

  gcc
  zstd

You may need to run `brew tap` to install additional taps.
```

**Comment by MikeMcQuaid** [comment]
Thanks @apainintheneck!
